prow/github/client.go

```diff
@@ -89,19 +90,36 @@ func (c *Client) log(methodName string, args ...interface{}) {
 	c.Logger.Printf("%s(%s)", methodName, strings.Join(as, ", "))
 }
 
-// Retry on transport failures. Does not retry on 500s.
+var timeSleep = time.Sleep
+
+// Retry on transport failures. Retries on 500s and retries after sleep on
+// ratelimit exceeded.
 func (c *Client) request(method, path string, body interface{}) (*http.Response, error) {
 	var resp *http.Response
 	var err error
 	backoff := retryDelay
 	for retries := 0; retries < maxRetries; retries++ {
 		resp, err = c.doRequest(method, path, body)
 		if err == nil {
-			break
+			// If we are out of API tokens, sleep first. The X-RateLimit-Reset
+			// header tells us the time at which we can request again.
+			if resp.StatusCode == 403 && resp.Header.Get("X-RateLimit-Remaining") == "0" {
+				resp.Body.Close()
+				var t int
+				if t, err = strconv.Atoi(resp.Header.Get("X-RateLimit-Reset")); err == nil {
+					timeSleep(time.Unix(int64(t), 0).Sub(time.Now()) + time.Second)
```

Comment:
From his commit message in https://github.com/kubernetes/test-infra/pull/640/commits/6a6217151b4f80a07852fc3eabcdd0cac5f7f0b1

> There is a "bug" in Github that they reset Remaining before Reset. Here
> are the log of it happening:
> 
> I0922 16:59:29.723383       1 token-counter.go:163] Processing token for 'k8s-merge-robot'
> I0922 16:59:29.777041       1 token-counter.go:123] k8s-merge-robot: Current rate: github.Rate{Limit:5000, Remaining:3053, Reset:github.Timestamp{2016-09-22 17:19:21 +0000 UTC}}. Sleeping for 9m55.611485s.
> I0922 17:09:25.661410       1 token-counter.go:123] k8s-merge-robot: Current rate: github.Rate{Limit:5000, Remaining:2417, Reset:github.Timestamp{2016-09-22 17:19:21 +0000 UTC}}. Sleeping for 4m57.669301663s.
> I0922 17:14:23.573919       1 token-counter.go:123] k8s-merge-robot: Current rate: github.Rate{Limit:5000, Remaining:2188, Reset:github.Timestamp{2016-09-22 17:19:21 +0000 UTC}}. Sleeping for 2m28.713046538s.
> I0922 17:16:52.480898       1 token-counter.go:123] k8s-merge-robot: Current rate: github.Rate{Limit:5000, Remaining:1958, Reset:github.Timestamp{2016-09-22 17:19:21 +0000 UTC}}. Sleeping for 1m14.259557755s.
> I0922 17:18:06.987068       1 token-counter.go:123] k8s-merge-robot: Current rate: github.Rate{Limit:5000, Remaining:1886, Reset:github.Timestamp{2016-09-22 17:19:21 +0000 UTC}}. Sleeping for 37.00647208s.
> I0922 17:18:44.190656       1 token-counter.go:123] k8s-merge-robot: Current rate: github.Rate{Limit:5000, Remaining:1856, Reset:github.Timestamp{2016-09-22 17:19:21 +0000 UTC}}. Sleeping for 18.404678296s.
> I0922 17:19:02.796591       1 token-counter.go:123] k8s-merge-robot: Current rate: github.Rate{Limit:5000, Remaining:1841, Reset:github.Timestamp{2016-09-22 17:19:21 +0000 UTC}}. Sleeping for 9.101710392s.
> I0922 17:19:11.952885       1 token-counter.go:123] k8s-merge-robot: Current rate: github.Rate{Limit:5000, Remaining:1832, Reset:github.Timestamp{2016-09-22 17:19:21 +0000 UTC}}. Sleeping for 4.523563543s.
> I0922 17:19:16.534085       1 token-counter.go:123] k8s-merge-robot: Current rate: github.Rate{Limit:5000, Remaining:1828, Reset:github.Timestamp{2016-09-22 17:19:21 +0000 UTC}}. Sleeping for 2.232963959s.
> I0922 17:19:18.822034       1 token-counter.go:123] k8s-merge-robot: Current rate: github.Rate{Limit:5000, Remaining:1826, Reset:github.Timestamp{2016-09-22 17:19:21 +0000 UTC}}. Sleeping for 1.088990686s.
> I0922 17:19:19.963956       1 token-counter.go:123] k8s-merge-robot: Current rate: github.Rate{Limit:5000, Remaining:1825, Reset:github.Timestamp{2016-09-22 17:19:21 +0000 UTC}}. Sleeping for 518.027818ms.
> I0922 17:19:20.536109       1 token-counter.go:123] k8s-merge-robot: Current rate: github.Rate{Limit:5000, Remaining:5000, Reset:github.Timestamp{2016-09-22 17:19:21 +0000 UTC}}. Sleeping for 231.952532ms.
> I0922 17:19:20.826210       1 token-counter.go:134] k8s-merge-robot: ### TOKEN USAGE: 0
> 
> As we can see in the last two fetch of current rate, they reset the
> value of Remaining back to 5000, but keep the Reset timestamp to the old
> value. Then they reset the timestamp value. This causes the program to
> log a token usage of 0.
> 
> Fix the bug by also resetting the counter when the Remaining value
> increases, and wait a little bit to make sure that the Reset value is
> also reset.